### PR TITLE
Set unique labels string on prometheus metrics

### PR DESCRIPTION
**Pull request by hamersaw**
# TL;DR
This PR computes a unique labels set for each metric on initialization. This is used to popular metric values during increments (so a new slice does not have to be constructed), but more importantly it fixes an issue with corrupting the global metric keys slice.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
The Golang `append` function implementation will reuse the underlying slice if it has remaining capacity. In the current implementation, this means that when we construct a new metric with additional labels we may modify the global metricKeys slice. This may result in the corruption of labels within other metrics.

## Tracking Issue
fixes https://github.com/flyteorg/flyte/issues/2597

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by hamersaw** [comment]
cc @honnix

**Comment by honnix** [comment]
Shall we add at least one test case to cover the corruption problem?

File: promutils/labeled/counter.go [hunk 1]
```diff
@@ -73,15 +61,19 @@ func NewCounter(name, description string, scope promutils.Scope, opts ...MetricO
...
 			c.Counter = scope.MustNewCounter(GetUnlabeledMetricName(name), description)
 		} else if additionalLabels, casted := opt.(AdditionalLabelsOption); casted {
-			labels := GetUniqueLabels(metricStringKeys, additionalLabels.Labels)
-			// Here we only append the labels that don't exist in metricStringKeys
-			c.CounterVec = scope.MustNewCounterVec(name, description, append(metricStringKeys, labels...)...)
-			c.additionalLabels = contextutils.MetricKeysFromStrings(labels)
+			// compute unique labels
+			labelSet := sets.NewString(metricStringKeys...)
+			labelSet.Insert(additionalLabels.Labels...)
+			labels := labelSet.List()
```
Comment (honnix) [inline promutils/labeled/counter.go:67]: This might not reserve the order of labels. I hope that doesn't matter.

File: promutils/labeled/counter.go [hunk 2]
```diff
@@ -73,15 +61,19 @@ func NewCounter(name, description string, scope promutils.Scope, opts ...MetricO
...
+
+			c.CounterVec = scope.MustNewCounterVec(name, description, labels...)
+			c.labels = contextutils.MetricKeysFromStrings(labels)
 		}
 	}
 
 	if c.CounterVec == nil {
 		c.CounterVec = scope.MustNewCounterVec(name, description, metricStringKeys...)
+		c.labels = metricKeys
```
Comment (honnix) [inline promutils/labeled/counter.go:76]: It might be good to cover this branch.

File: promutils/labeled/gauge.go [hunk 3]
```diff
@@ -113,14 +115,19 @@ func NewGauge(name, description string, scope promutils.Scope, opts ...MetricOpt
...
+			labels := labelSet.List()
+
+			g.GaugeVec = scope.MustNewGaugeVec(name, description, labels...)
+			g.labels = contextutils.MetricKeysFromStrings(labels)
 		}
 	}
 
 	if g.GaugeVec == nil {
 		g.GaugeVec = scope.MustNewGaugeVec(name, description, metricStringKeys...)
+		g.labels = metricKeys
```
Comment (honnix) [inline promutils/labeled/gauge.go:130]: It might be good to cover this branch.

**Review by EngHabu — DISMISSED** [review]
Thank you for fixing this!

File: promutils/labeled/counter.go (hunk 2 above)
Comment (hamersaw) [reply promutils/labeled/counter.go:76]: New tests add this coverage.

File: promutils/labeled/gauge.go (hunk 3 above)
Comment (hamersaw) [reply promutils/labeled/gauge.go:130]: New tests add this coverage.

File: promutils/labeled/counter.go (hunk 1 above)
Comment (hamersaw) [reply promutils/labeled/counter.go:67]: Internally these labels [are sorted](https://github.com/prometheus/client_golang/blob/2c3d072cdd4dbd9dbbfe415cbd01851db8962289/prometheus/desc.go#L106) before stored, so this shouldn't be an issue. If it ever changes, it will break the unit new unit tests so I think we are OK here.

Comment (honnix) [reply promutils/labeled/counter.go:67]: Great! Thanks for the pointer.

**Review by honnix — APPROVED** [review]

**Comment by honnix** [comment]
Shall we get this merged and upgrade related components?

**Review by EngHabu — APPROVED** [review]
